### PR TITLE
Add dynamic configuration of Holidays

### PR DESCRIPTION
**Pull request by webecke**
## Overview

Refactored holiday storage to use the configuration system.

## Details

- Removed the hard-coded list of holidays
- Added configuration for holidays
- Added front-end warning for when the list of holidays doesn't go more than 30 days into the future (see screenshot)

image

## Testing

- [ ] Added/updated unit tests
- [ ] Tested edge cases
- [X] Manual testing (if needed)

## Dependencies

This relies on the configuration refactoring work done in #524, which is why this PR goes to `config-fullstack-refactor` and not `main`

## Additional Notes

>[!CAUTION]
>As with all most configuration updates, all phases on the live Autograder will need to be disabled temporarily when this update goes to production. Then someone will have to go in and add all the holidays. Otherwise the Autograder won't know when holidays are.

### CODE REVIEW

**Comment by webecke** [comment]
Yes I'm aware this can merge already since its not going to main. However, I'd like this to be a separate PR with its own code review and discussion since it is a fundamentally different update.

**Review by 19mdavenport — APPROVED** [review]
Looks pretty good to me, did you do any manual testing to make sure the holidays with this system were counted?

**Comment by frozenfrank** [comment]
> Looks pretty good to me, did you do any manual testing to make sure the holidays with this system were counted correctly while grading?

The holiday system already has extensive tests created for it. Previously, the system was reading the config value from a hard-coded string, and this PR changes that string to come from a dynamic location instead.

**Comment by 19mdavenport** [comment]
Fair, I would still like an answer though.

**Comment by webecke** [comment]
I did some very basic manual testing to make sure they worked. Just enough to make sure the dates were being loaded correctly into the `LateDayCalculator` but nothing extensive. Lots of errors get thrown if my system breaks, though I can't think of a scenario where it would break, there's several data validation steps to make sure dates are valid and whatnot.

**Comment by webecke** [comment]
Oh would you look at that, `LateDayCalculator` tests are failing on empty now, not sure what that's happening. I'll have to investigate that tomorrow or next week.

File: src/main/resources/frontend/src/components/config/HolidayConfigEditor.vue [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  holidaysSet.value = new Set();
+};
+
+const submitHolidays = async () => {
+  try {
+    await updateHolidays([...holidaysSet.value]);
+    closeEditor();
+  } catch (e) {
+    useConfigStore().updatePublicConfig();
+    alert("There was a problem while saving holidays");
```
Comment (frozenfrank) [inline src/main/resources/frontend/src/components/config/HolidayConfigEditor.vue:52]: Good job with the error handling!

File: src/main/resources/frontend/src/components/config/HolidayConfigEditor.vue [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    <input type="date" v-model="currentDateToAdd" />
+    <button class="small" :disabled="!currentDateToAdd" @click="addHolidayToList">
+      <i class="fa-solid fa-plus" />
+    </button>
+  </div>
+  <button @click="submitHolidays" style="margin-top: 10px">Submit</button>
+  <Panel style="max-width: 400px">
+    <p>
+      Holidays don't count towards the late-day count for submissions. The Autograder already treats
+      every Saturday and Sunday as a "holiday", so it is only necessary to add BYU specific
```
Comment (frozenfrank) [inline src/main/resources/frontend/src/components/config/HolidayConfigEditor.vue:79]: Is it worth including a link to the official academic calendar here?

https://academiccalendar.byu.edu/

File: src/main/resources/frontend/src/utils/utils.ts [hunk 3]
```diff
@@ -251,3 +244,30 @@ export const resultsScoreDisplayText = (results: RubricItemResults): string => {
...
+  if (!dates.length) return true;
+
+  let datesToUse: Date[];
+  if (typeof dates[0] === "string") {
+    datesToUse = dates.map((d) => new Date(d as string));
+  } else {
+    datesToUse = dates as Date[];
+  }
+
+  datesToUse.sort((a: Date, b: Date) => a.getTime() - b.getTime());
```
Comment (frozenfrank) [inline src/main/resources/frontend/src/utils/utils.ts:264]: Sorting the list isn't necessary here. Sorting takes `O(n * logn)` time, but we only need to get the latest date. That only needs to require an `O(n)` sweep through the data.

File: src/main/resources/frontend/src/utils/utils.ts [hunk 4]
```diff
@@ -251,3 +244,30 @@ export const resultsScoreDisplayText = (results: RubricItemResults): string => {
...
+  const lastHoliday = datesToUse[datesToUse.length - 1];
+  const today = new Date();
+  today.setHours(0, 0, 0, 0);
+  lastHoliday.setHours(0, 0, 0, 0);
+
+  const daysInTheFuture = (lastHoliday.getTime() - today.getTime()) / (1000 * 60 * 60 * 24);
```
Comment (frozenfrank) [inline src/main/resources/frontend/src/utils/utils.ts:270]: This is the kind of date arithematic that is probably 99.9% correct right now, but also potentially ripe with unhandled edge cases. Do you think it's worth switching to a date library like [MomentJS](https://momentjs.com/)?

File: src/main/resources/frontend/src/views/AdminView/ConfigView.vue [hunk 5]
```diff
@@ -101,6 +111,26 @@ onMounted(async () => {
...
+    >
+      <template #editor="{ closeEditor }">
+        <HolidayConfigEditor :closeEditor="closeEditor" />
+      </template>
+      <template #current>
+        <div v-if="holidayWarning()">
+          <b style="background-color: red; color: white; border-radius: 5px; padding: 5px"
+            >Holidays are about to run out!</b
+          >
+          <p><em>Please add more holidays prompty, using the University Academic Calendar.</em></p>
```
Comment (frozenfrank) [inline src/main/resources/frontend/src/views/AdminView/ConfigView.vue:126]: Should this also have a direct link to the university calendar?

**Review by frozenfrank — COMMENTED** [review]
These changes look very nice. Thanks for implementing this!

I left comments about a few decisions represented in the code. Some of those may be worth looking into, but they aren't intended to be show stoppers.

I'll give the test cases a look over to see if I can help get them to pass.

**Comment by frozenfrank** [comment]
> Oh would you look at that, `LateDayCalculator` tests are failing on empty now, not sure what that's happening. I'll have to investigate that tomorrow or next week.

@webecke I updated the date formats so that the `LateDayCalculatorTest` and _most_ of the `ScorerTest` tests pass.

However, three tests are still failing, and it's not obviously related to the date formatting issue. I'm not sure what's going on here. Can someone help figure out what might be causing these tests to fail? The failing tests are all related to fancy score decreasing behaviors.

image

File: src/main/java/edu/byu/cs/autograder/score/LateDayCalculator.java [hunk 6]
```diff
@@ -38,14 +38,23 @@
...
 public class LateDayCalculator {
 
+    private static final String DEFAULT_DATE_FORMAT = "yyyy-MM-dd";
     private static final Logger LOGGER = Logger.getLogger(LateDayCalculator.class.getName());
 
     private Set<LocalDate> publicHolidays;
     private final Map<String, LateDayContext> lateDayContextCache = new HashMap<>();
 
     public LateDayCalculator() {
+        System.out.println("Initializing standard LateDayCalculator");
```
Comment (frozenfrank) [inline src/main/java/edu/byu/cs/autograder/score/LateDayCalculator.java:48]: Oops 🙊 I didn't mean to leave that in there. I should have been watching my commits more closely.

File: src/main/java/edu/byu/cs/autograder/score/LateDayCalculator.java [hunk 7]
```diff
@@ -434,8 +443,13 @@ private boolean isPublicHoliday(@NonNull ZonedDateTime zonedDateTime) {
     }
 
     private String getEncodedPublicHolidays() {
-        // FIXME: Return from some dynamic location like a configuration file or a configurable table
-        return "1/1/2025;1/20/2025;2/17/2025;3/21/2025;5/26/2025;6/19/2025;7/4/2025;7/24/2025;9/1/2025;11/26/2025;11/27/2025;11/28/2025;12/25/2025;12/26/2025;"
-                + "1/1/2026;";
+        String encodedHolidays;
+        try {
+            encodedHolidays = DaoService.getConfigurationDao().getConfiguration(ConfigurationDao.Configuration.HOLIDAY_LIST, String.class);
```
Comment (frozenfrank) [inline src/main/java/edu/byu/cs/autograder/score/LateDayCalculator.java:448]: I'd rather just return it directly, and maybe split up this long line a bit.

**Review by frozenfrank — COMMENTED** [review]
Besides the failing test cases, I have some minor clean up work to do before merging this.

File: src/main/java/edu/byu/cs/autograder/score/LateDayCalculator.java [hunk 8]
```diff
@@ -250,14 +250,14 @@ public Set<LocalDate> initializePublicHolidays(@NonNull String encodedPublicHoli
...
-     * that will accept strings matching this example: "9/16/2024"
+     * that will accept strings matching this example: "2025-12-25"
      *
      * @see #initializePublicHolidays(String, String, boolean)
      *
      * @param encodedPublicHolidays A string representing the encoded data.
      */
     public Set<LocalDate> initializePublicHolidays(@NonNull String encodedPublicHolidays, boolean quietWarnings) {
-        return initializePublicHolidays(encodedPublicHolidays, "M/d/yyyy", quietWarnings);
+        return initializePublicHolidays(encodedPublicHolidays, "yyyy-MM-dd", quietWarnings);
```
[the remark this answers is not in the document]
Comment (webecke) [reply src/main/java/edu/byu/cs/autograder/score/LateDayCalculator.java]: We decided `yyyy-mm-dd`

File: src/main/java/edu/byu/cs/service/ConfigService.java [hunk 9]
```diff
@@ -133,6 +116,27 @@ public static ArrayList<PrivateConfig.AssignmentConfig> generateAssignmentsConfi
         return assignments;
     }
 
+    public static String[] generateHolidayConfig() throws DataAccessException {
```
[the remark this answers is not in the document]
Comment (webecke) [reply src/main/java/edu/byu/cs/service/ConfigService.java:119]: Well, the other similar functions are called "generateX". We're generating the object for the frontend

File: src/main/resources/frontend/src/utils/utils.ts (hunk 4 above)
Comment (webecke) [reply src/main/resources/frontend/src/utils/utils.ts:270]: When is this not going to work? Daylight savings and be off by one hour? I don't see a point in adding more dependencies

File: src/main/resources/frontend/src/views/AdminView/ConfigView.vue (hunk 5 above)
Comment (webecke) [reply src/main/resources/frontend/src/views/AdminView/ConfigView.vue:126]: No, I'll add it to the edit screen.

File: src/main/resources/frontend/src/utils/utils.ts (hunk 3 above)
Comment (webecke) [reply src/main/resources/frontend/src/utils/utils.ts:264]: good point

**Comment by webecke** [comment]
Wow this is fun, we can't merge any PRs for the rest of the month

**Comment by 19mdavenport** [comment]
> Wow this is fun, we can't merge any PRs for the rest of the month

Right now I think it may be related to storage space rather than minutes

**Comment by 19mdavenport** [comment]
> Wow this is fun, we can't merge any PRs for the rest of the month

Also I can override the check if I need to. What I'm more worried about is not being able to build it

**Comment by frozenfrank** [comment]
> Right now I think it may be related to storage space rather than minutes

@19mdavenport This sounds feasible. The total storage limit is 500MB.

I am able to see our actions minutes usage, but it appears that viewing the storage usage requires more permissions. Here is the information about how to find the storage usage: [Viewing your GitHub Actions Usage](https://docs.github.com/en/billing/managing-billing-for-your-products/managing-billing-for-github-actions/viewing-your-github-actions-usage)

**Comment by webecke** [comment]
@frozenfrank I'm trying to close out this PR and it seems that something you changed on main changed ScorerTest in an unexpected way, so now I have a merge conflict I can't figure out. Any clue what's going on?

**Comment by frozenfrank** [comment]
> @frozenfrank I'm trying to close out this PR and it seems that something you changed on main changed ScorerTest in an unexpected way, so now I have a merge conflict I can't figure out. Any clue what's going on?

@webecke I'll look at this shortly

**Comment by frozenfrank** [comment]
> @frozenfrank I'm trying to close out this PR and it seems that something you changed on main changed ScorerTest in an unexpected way, so now I have a merge conflict I can't figure out. Any clue what's going on?

@webecke Merge conflicts resolved. Merge now!!

Merge now!

**Comment by webecke** [comment]
> @webecke Merge conflicts resolved. Merge now!!

Huzzah! Thank you! I'm gonna do some quick manual testing to make sure nothing broke then merge this PR

**Comment by webecke** [comment]
@19mdavenport, now that this is merged, you must remember to deactivate graded phases the next time you make a new version and deploy the autograder. Holidays will not be configured until someone goes in and adds them manually.
